blockchain/sync_manager/src/sync.rs

```diff
@@ -122,6 +146,141 @@ impl<'a> Syncer<'a> {
+                        }
+                    } else {
+                        return Err(Error::State(
+                            "Could not retrieve actor from state tree".to_string(),
+                        ));
+                    }
+                }
+            };
+            msg_meta_data
+                .metadata
+                .insert(msg.from().clone(), updated_state);
+            Ok(())
+        }
+
+        let mut msg_meta_data = MsgCheck::new();
+        // TODO retrieve tipset state and load state tree
+        // temporary
+        let tree = HamtStateTree::default();
+        // loop through bls messages and check msg validity
+        for m in block.bls_msgs() {
+            check_msg(m.clone(), &mut msg_meta_data, &tree)?;
+        }
+        // loop through secp messages and check msg validity and signature
+        for m in block.secp_msgs() {
+            check_msg(m.clone(), &mut msg_meta_data, &tree)?;
+            // signature validation
+            if !is_valid_signature(&m.cid()?.to_bytes(), m.from(), m.signature()) {
+                return Err(Error::Validation(
+                    "Message signature is not valid".to_string(),
+                ));
+            }
+        }
+        // validate message root from header matches message root
+        let sm_root = self.compute_msg_data(&block)?;
+        if block.to_header().messages() != &sm_root {
+            return Err(Error::InvalidRoots);
+        }
+
+        Ok(())
+    }
+
+    /// Validates block semantically according to https://github.com/filecoin-project/specs/blob/6ab401c0b92efb6420c6e198ec387cf56dc86057/validation.md
+    pub fn validate(&self, block: Block) -> Result<(), Error> {
+        // get header from full block
+        let header = block.to_header();
+        let base_tipset = self.load_fts(TipSetKeys::new((*header.parents().cids).to_vec()))?;
+
+        // check if block has been signed
+        if header.signature().bytes().is_empty() {
+            return Err(Error::Validation("Signature is nil in header".to_string()));
+        }
+
+        // first check that it is not in the future; see https://github.com/filecoin-project/specs/blob/6ab401c0b92efb6420c6e198ec387cf56dc86057/validation.md
+        // allowing for some small grace period to deal with small asynchrony
+        // using allowable_clock_drift from Lotus; see https://github.com/filecoin-project/lotus/blob/master/build/params_shared.go#L34:7
+        let allowable_clock_drift: u64 = 1;
+        let time_now = SystemTime::now().duration_since(UNIX_EPOCH)?;
+        if header.timestamp() > time_now.as_secs() + allowable_clock_drift {
+            return Err(Error::Validation("Block was from the future".to_string()));
+        }
+        if header.timestamp() > time_now.as_secs() {
+            return Err(Error::Validation(
+                "Got block from the future, but within threshold".to_string(),
+            ));
+        }
```

Comment:
I'm confused, why do we need to handle these seperately if they return the same error variant?